pkg/config/generate.go

```diff
@@ -206,114 +202,171 @@ func Generate(m *api.Manifest) (c *Config, err error) {
 	for _, cert := range certs {
 		if cert.signingKey == nil && cert.signingCert == nil {
 			cert.signingKey, cert.signingCert = c.CaKey, c.CaCert
 		}
+		if *cert.key != nil && *cert.cert != nil &&
+			(*cert.cert).CheckSignatureFrom(cert.signingCert) == nil {
+			continue
+		}
 		if *cert.key, *cert.cert, err = tls.NewCert(cert.cn, cert.organization, cert.dnsNames, cert.ipAddresses, cert.extKeyUsage, cert.signingKey, cert.signingCert); err != nil {
 			return
 		}
 	}
 
-	if c.ServiceAccountPrivateKey, err = tls.NewPrivateKey(); err != nil {
-		return
+	secrets := []struct {
+		secret *[]byte
+		n      int
+	}{
+		{
+			secret: &c.SessionSecretAuth,
+			n:      24,
+		},
+		{
+			secret: &c.SessionSecretEnc,
+			n:      24,
+		},
+		{
+			secret: &c.RegistryHTTPSecret,
+		},
+		{
+			secret: &c.AlertManagerProxySessionSecret,
+		},
+		{
+			secret: &c.AlertsProxySessionSecret,
+		},
+		{
+			secret: &c.PrometheusProxySessionSecret,
+		},
+	}
+	for _, s := range secrets {
+		if len(*s.secret) != 0 {
+			continue
+		}
+		if s.n == 0 {
+			s.n = 32
+		}
+		if *s.secret, err = randomBytes(s.n); err != nil {
+			return
+		}
 	}
-	c.ServiceAccountPublicKey = &c.ServiceAccountPrivateKey.PublicKey
 
-	if c.SessionSecretAuth, err = randomBytes(24); err != nil {
-		return
+	kubeconfigs := []struct {
+		clientKey  *rsa.PrivateKey
+		clientCert *x509.Certificate
+		endpoint   string
+		username   string
+		namespace  string
+		kubeconfig **v1.Config
+	}{
+		{
+			clientKey:  c.OpenShiftMasterKey,
+			clientCert: c.OpenShiftMasterCert,
+			endpoint:   "master-api",
+			username:   "system:openshift-master",
+			kubeconfig: &c.MasterKubeconfig,
+		},
+		{
+			clientKey:  c.ServiceCatalogAPIClientKey,
+			clientCert: c.ServiceCatalogAPIClientCert,
+			endpoint:   "master-api",
+			username:   "system:serviceaccount:kube-service-catalog:service-catalog-apiserver",
+			namespace:  "kube-service-catalog",
+			kubeconfig: &c.ServiceCatalogAPIKubeconfig,
+		},
+		{
+			clientKey:  c.BootstrapAutoapproverKey,
+			clientCert: c.BootstrapAutoapproverCert,
+			endpoint:   "master-api",
+			username:   "system:serviceaccount:openshift-infra:bootstrap-autoapprover",
+			namespace:  "openshift-infra",
+			kubeconfig: &c.BootstrapAutoapproverKubeconfig,
+		},
+		{
+			clientKey:  c.AdminKey,
+			clientCert: c.AdminCert,
+			endpoint:   m.PublicHostname,
+			username:   "system:admin",
+			kubeconfig: &c.AdminKubeconfig,
+		},
+		{
+			clientKey:  c.NodeBootstrapKey,
+			clientCert: c.NodeBootstrapCert,
+			endpoint:   m.PublicHostname,
+			username:   "system:serviceaccount:openshift-infra:node-bootstrapper",
+			kubeconfig: &c.NodeBootstrapKubeconfig,
+		},
 	}
-	if c.SessionSecretEnc, err = randomBytes(24); err != nil {
-		return
+	for _, kc := range kubeconfigs {
+		if kc.namespace == "" {
+			kc.namespace = "default"
+		}
+		if *kc.kubeconfig, err = makeKubeConfig(kc.clientKey, kc.clientCert, c.CaCert, kc.endpoint, kc.username, kc.namespace); err != nil {
+			return
```

Comment:
for other reviewers, the public interface method ensures nil is returned if an error is present, not a partially defined config.